### DISCUSSION TITLE
I am learning EGSnrc but having difficulties with Getting-started.pdf

### DISCUSSION THREAD

**Discussion by kstcmcjw** [nrc-cnrc/EGSnrc#814] [Questions] [answered]
Hello everyone:
    i am a new learner of EGSnrc, i learned the Getting-started.pdf of myself,  the chapter 5: 4.1 Create a new egs++ application, the smallest, fully functional EGSnrc application, I can't compile , this is the document of my computer:
![image](https://user-images.githubusercontent.com/96934217/147853334-2c8819e0-bb5a-4770-ba07-65de84693f49.png)

duo to my slow-wittied, the chapter 4 teach me compile first easy egs++ application, and a pencil beam on tantalum, or on lead, or on water. The example simulation is very perfect, but I don't know how to use this EGS, please allow me to describe my process.
 
First, creating a new directory for your application, myapp. It is under the folder of : C:\EGSnrc\EGS_HOME. 
Second, copy the two files "array_sizes.h, Makefile" from tutor7pp to "myapp" folder, and open the "Makefile", change the "tutor7pp" to "myapp".(USER_CODE = myapp   and   EGSPP_USER_MACROS = myapp.macros)
Third, created a txt, change its name to "myapp.macros", and this is an empty file.
Fourth, created a txt, change its name to "myapp.cpp", and type the code "#include "egs_advanced_application.h" "and "APP_MAIN (EGS_AdvancedApplication)".

OK, there are four files now, and i open the EGS_gui, and compiled it, it shows exticode=2, not 0, why ? I am sure i go with the chapter 4.1, where is my mistake?

I realy want to know, thank you for your criticism, and wish you happy new year!
![image](https://user-images.githubusercontent.com/96934217/147853355-daca3a48-86cc-47fe-81ff-70f84cf45f93.png)

Thanks for your answer, happy 2022!

**Comment by crcrewso** [comment]
Have you looked at Chapter 4, and Tutorial 7?

**Reply by kstcmcjw** [reply]
where is Tutorial 7？ And i reply a email to you! thank U very much!

**Comment by Jaap1957** [comment]
You skipped a step in the getting-started.pdf, 

If you open a shell at the folder C:\EGSnrc\EGS_HOME\myapp
and give the command 

`C:\EGSnrc\EGS_HOME\myapp>make`

Or

`C:\EGSnrc\EGS_HOME\myapp>mingw32-make`

what happens then?

**Reply by kstcmcjw** [reply]
> You skipped a step in the getting-started.pdf,
> 
> If you open a shell at the folder C:\EGSnrc\EGS_HOME\myapp and give the command
> 
> `C:\EGSnrc\EGS_HOME\myapp>make`
> 
> Or
> 
> `C:\EGSnrc\EGS_HOME\myapp>mingw32-make`
> 
> what happens then?

Thanks for your reply,  i  really miss the "make" function indeed, i thought this step is on linux system, and i am win10 system, so i didn't do the "make" step.  i will have a try today with the mingw32-make.


![image](https://user-images.githubusercontent.com/96934217/148003943-fc563a86-7875-4ef1-8f89-bed385c51937.png)

And i find another problem my flex edition is not 2.6.4, may lead the exitcode=2, i will update and have a try presistent.

**Comment by Jaap1957** [comment]
Here a few remarks i noted down during my queeste towards EGSnrc working.

Start here
[https://github.com/nrc-cnrc/EGSnrc/wiki/install-EGSnrc-on-windows](url)

Download and install GIT to make sure you do not change anything you will be sorry for afterwards.

Download / install mingw version 7.3.0

Make sure the path to g++ / gcc / gfortran is in your environment variables.
To check/change:
Go to Control Panel\All Control Panel Items\System
in left pane click Advanced system settings
tab Advanced
click Envrionment Variables
either in System varibles or in User variables click Path
add 
c:\Program Files\mingw-w64\x86_64-7.3.0-posix-seh-rt_v5-rev0\mingw64\bin\
to the list. (please check this is indeed the path to mingw at your system)

Download EGSnrc-configure-windows.exe  at
[https://github.com/nrc-cnrc/EGSnrc/releases](url)

Make sure to put EGSnrc-configure-windows.exe  
IN THIS PATH:
\EGSNRC\EGSnrc-configure-windows.exe 

ottherwise it will not work.

I got a missing libiconv-2.dll. I could copy it from 
"c:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\CommonExtensions\Microsoft\TeamFoundation\Team Explorer\Git\mingw32\bin\libiconv-2.dll" 
to
"c:\Program Files\mingw-w64\x86_64-7.3.0-posix-seh-rt_v5-rev0\mingw64\bin\.." 

I do not know your application, but I did not need TCL or QTgrace up to now.

So next step is what you already did, get the 4 files in 'myapp', as described in getting-started.pdf

start mingw32-make in the myapp folder
fingers crossed
Your executable should show up in 
\EGSnrc\EGS_home\bin\win3264\myapp.exe

Do not go there ... just stay in the myapp folder.
\EGSnrc\EGS_huisje\myapp\myapp.exe -i slab.egsinp
should work now

And then, the fun part: viewing tracks
Run
\EGSnrc\EGS_home\myapp\egs_view slab.egsinp slab.ptracks

**Reply by kstcmcjw** [reply]
> Here a few remarks i noted down during my queeste towards EGSnrc working.
> 
> Start here [https://github.com/nrc-cnrc/EGSnrc/wiki/install-EGSnrc-on-windows](url)
> 
> Download and install GIT to make sure you do not change anything you will be sorry for afterwards.
> 
> Download / install mingw version 7.3.0
> 
> Make sure the path to g++ / gcc / gfortran is in your environment variables. To check/change: Go to Control Panel\All Control Panel Items\System in left pane click Advanced system settings tab Advanced click Envrionment Variables either in System varibles or in User variables click Path add c:\Program Files\mingw-w64\x86_64-7.3.0-posix-seh-rt_v5-rev0\mingw64\bin to the list. (please check this is indeed the path to mingw at your system)
> 
> Download EGSnrc-configure-windows.exe at [https://github.com/nrc-cnrc/EGSnrc/releases](url)
> 
> Make sure to put EGSnrc-configure-windows.exe IN THIS PATH: \EGSNRC\EGSnrc-configure-windows.exe
> 
> ottherwise it will not work.
> 
> I got a missing libiconv-2.dll. I could copy it from "c:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\CommonExtensions\Microsoft\TeamFoundation\Team Explorer\Git\mingw32\bin\libiconv-2.dll" to "c:\Program Files\mingw-w64\x86_64-7.3.0-posix-seh-rt_v5-rev0\mingw64\bin.."
> 
> I do not know your application, but I did not need TCL or QTgrace up to now.
> 
> So next step is what you already did, get the 4 files in 'myapp', as described in getting-started.pdf
> 
> start mingw32-make in the myapp folder fingers crossed Your executable should show up in \EGSnrc\EGS_home\bin\win3264\myapp.exe
> 
> Do not go there ... just stay in the myapp folder. \EGSnrc\EGS_huisje\myapp\myapp.exe -i slab.egsinp should work now
> 
> And then, the fun part: viewing tracks Run \EGSnrc\EGS_home\myapp\egs_view slab.egsinp slab.ptracks

Thank you for your very detailed reply, i find some place worthing recheck: 1.As you said: click Envrionment Variables--either in System varibles or in User variables click Path, i only add the System varibles except  User variables, 2.  add c:\Program Files\mingw-w64\x86_64-7.3.0-posix-seh-rt_v5-rev0\mingw64\bin to the list. (please check this is indeed the path to mingw at your system)   i  think i may miss this path.

Apologise for my  stupid, how can i do "start mingw32-make in the myapp folder"?   And  mingw32-make is what? I install MinGW-w64 and the teminal is this ? But it can't be work,  why?  Can you give me a picture please?

![image](https://user-images.githubusercontent.com/96934217/148022109-5000e74f-b903-448d-a76b-3f5b2fcf1b00.png)
![image](https://user-images.githubusercontent.com/96934217/148022672-a3bbdb2e-057b-44e5-b30d-386822dda6ed.png)


why the EGS_gui using guide don't mention   'make'? 
![image](https://user-images.githubusercontent.com/96934217/148022359-7ecf952e-3812-4ba9-955a-beffb10f60a8.png)

**Comment by Jaap1957** [comment, marked as answer]
> how can i do "start mingw32-make in the myapp folder"? 

Open a command window, go to the myapp folder like this screenshot:
![afbeelding](https://user-images.githubusercontent.com/96918927/148027213-840e9453-51c7-4e11-9e99-62ac5a166854.png)

As you see the 4 files are there ... like yours
Assuming paths are set correct then type the command at the command prompt:
At my PC i then get ...
![afbeelding](https://user-images.githubusercontent.com/96918927/148027793-0381c9c3-02e2-4ded-9f6c-3fd596218e69.png)

It takes a minute or 2 to fill your screen with the blabla.

> And mingw32-make is what? I install MinGW-w64 and the teminal is this ? But it can't be work, why? Can you give me a picture please?

mingw32-make can be found at

"c:\Program Files\mingw-w64\x86_64-7.3.0-posix-seh-rt_v5-rev0\mingw64\bin\mingw32-make.exe" 

and when you succesfully added the path
"c:\Program Files\mingw-w64\x86_64-7.3.0-posix-seh-rt_v5-rev0\mingw64\bin\" 
to the environment variable PATH, it should work like in the above screenshot.

> why the EGS_gui using guide don't mention 'make'?

When you have gone through the install process completely, egs-gui will do all the stuff.
But you are not at that point yet.

Keep going ! almost done ...

**Reply by kstcmcjw** [reply]
> > 我该如何“在 myapp 文件夹中启动 mingw32-make”？
> 
> 打开命令窗口，转到 myapp 文件夹，如下图所示： ![深信不疑](https://user-images.githubusercontent.com/96918927/148027213-840e9453-51c7-4e11-9e99-62ac5a166854.png)
> 
> 如您所见，有 4 个文件......就像你的一样 假设路径设置正确，然后在命令提示符下键入命令： 在我的 PC 上，我得到...... ![深信不疑](https://user-images.githubusercontent.com/96918927/148027793-0381c9c3-02e2-4ded-9f6c-3fd596218e69.png)
> 
> 用 blabla 填满你的屏幕需要一两分钟。
> 
> > 而mingw32-make是什么？我安装了 MinGW-w64，终端是这个吗？但它不能工作，为什么？你能给我一张照片吗？
> 
> mingw32-make 可以在
> 
> “c:\Program Files\mingw-w64\x86_64-7.3.0-posix-seh-rt_v5-rev0\mingw64\bin\mingw32-make.exe”
> 
> 当您成功地将路径 “c:\Program Files\mingw-w64\x86_64-7.3.0-posix-seh-rt_v5-rev0\mingw64\bin”添加 到环境变量PATH时，它应该像上面的截图一样工作.
> 
> > 为什么 EGS_gui 使用指南没有提到“制作”？
> 
> 当您完全完成安装过程后，egs-gui 将完成所有工作。 但你还没有到那个时候。
> 
> 继续前进 ！快完成了...

Oh, surpirse! It's very kind of you for answer my question, i got it, and i will have a try to achieve the  victory! I will study EGSnrc very earnest as your and some friends' very enthusiastic answer!  It was a great inspiration to me!  Thank you very much !

**Comment by kstcmcjw** [comment]
Thanks all the friends who has helped me @Jaap1957 @crcrewso . I had already successfully running EGS_gui without error code, indeed i forgot to add the PATH of GNU MAKE--x86_64-posix-seh, and after add path , cmd.exe  program runs OK, here is the picture:
![image](https://user-images.githubusercontent.com/96934217/148182277-1688bb11-0cc3-41b4-8456-6bfd59280f88.png)
Then , run the example of chapter 4.2 success, and simulation the figure:
![image](https://user-images.githubusercontent.com/96934217/148182573-b9c406b7-7cb7-4210-a8bc-1d3fa61a13e1.png)
And i will help the others who need helped, thanks to all the friends!

**Comment by kstcmcjw** [comment]
Thanks all the friends who has helped me @Jaap1957 @crcrewso . I had already successfully running EGS_gui without error code, indeed i forgot to add the PATH of GNU MAKE--x86_64-posix-seh, and after add path , cmd.exe  program runs OK, here is the picture:
![image](https://user-images.githubusercontent.com/96934217/148182277-1688bb11-0cc3-41b4-8456-6bfd59280f88.png)
Then , run the example of chapter 4.2 success, and simulation the figure:
![image](https://user-images.githubusercontent.com/96934217/148182573-b9c406b7-7cb7-4210-a8bc-1d3fa61a13e1.png)
And i will help the others who need be helped, thanks to all the friends!

**Reply by Mona99-del** [reply]
can you help because I have the same problem?

**Reply by rtownson** [reply]
@Mona99-del See the solution for the quadmath error here:

https://github.com/nrc-cnrc/EGSnrc/pull/972#issuecomment-1458833983

It will be fixed in the next release.

**Comment by Mona99-del** [comment]
Thanks for your reply
l tried to compile turtor7pp and successful but when tried run the myapp
after doing the steps in gutting start manual the error ming32-make.**no
rule to make target'myapp.macros',needed by
'egsnrc_win3264.F'.stop.ExitCode=2and after preparation slab.egsinp and
used in egs_view it successful and can see slab,I wrote the command myapp
-i slab.egsinp but the appear (Unsupported 16-Bit Application

The program or feature

"\??\C:\Users\dell\EGSnrc\EGS_Home\myapp\myapp.exe" cannot start or run due
to incompatibity with 64-bit versions of Windows. Please contact the
software vendor to ask if a 64-bit Windows compatible version is
available.)can know what the reason? Knowing when trying to run turtor7pp
is run and the  file ptrack apper and can display  in egs view

**Reply by rtownson** [reply]
In section 4.1 did you do the step "Create a new file called myapp.macros , and just save it as an empty file."?

**Reply by Mona99-del** [reply]
yes

**Reply by Mona99-del** [reply]
![image](https://github.com/user-attachments/assets/91eef803-04ad-4fbd-9d64-af41d8db5cf6)
this files

**Reply by Mona99-del** [reply]
![egs view slab](https://github.com/user-attachments/assets/9581f4e5-0973-43f4-94b6-615d6908646b)

**Reply by rtownson** [reply]
It says that myapp.macros is a text file, which tells me the extension is probably wrong. Is the extension actually .txt? Make sure you turn on viewing of extensions so that you can double check this.

**Comment by Mona99-del** [comment]
Yes here is the error l created by notepad
